kernel/src/device/char.rs

```diff
@@ -0,0 +1,212 @@
+pub fn acquire_major(major: MajorId) -> Result<MajorIdOwner> {
+    if major.get() == 0 || major.get() > MAX_MAJOR {
+        return_errno_with_message!(Errno::EINVAL, "invalid major ID");
+    }
+
+    if MAJORS.lock().insert(major.get()) {
+        Ok(MajorIdOwner(major))
+    } else {
+        return_errno_with_message!(Errno::EEXIST, "major ID already acquired")
+    }
+}
+
+/// Allocates a major ID.
+///
+/// The returned `MajorIdOwner` object represents the ownership to the major ID.
+/// Until the object is dropped, this major ID cannot be acquired via `acquire_major` or `allocate_major` again.
+pub fn allocate_major() -> Result<MajorIdOwner> {
+    let mut majors = MAJORS.lock();
+
+    for id in (CHAR_FIRST_DYNAMIC_MAJOR_START..CHAR_FIRST_DYNAMIC_MAJOR_END + 1).rev() {
+        if majors.insert(id) {
+            return Ok(MajorIdOwner(MajorId::new(id)));
+        }
+    }
+
+    for id in (CHAR_SECOND_DYNAMIC_MAJOR_START..CHAR_SECOND_DYNAMIC_MAJOR_END + 1).rev() {
+        if majors.insert(id) {
+            return Ok(MajorIdOwner(MajorId::new(id)));
+        }
+    }
+
+    return_errno_with_message!(Errno::ENOSPC, "no more major IDs available");
+}
+
+/// An owned major ID.
+///
+/// Each instances of this type will unregister the major ID when dropped.
+pub struct MajorIdOwner(MajorId);
+
+impl MajorIdOwner {
+    /// Returns the major ID.
+    pub fn get(&self) -> MajorId {
+        self.0
+    }
+}
+
+impl Drop for MajorIdOwner {
+    fn drop(&mut self) {
+        MAJORS.lock().remove(&self.0.get());
+    }
+}
+
+pub(super) fn init_in_first_process(fs_resolver: &FsResolver) -> Result<()> {
+    for device in collect_all() {
+        let name = device.name().to_string();
+        let device = Arc::new(CharFile::new(device));
+        add_node(device, &name, fs_resolver)?;
+    }
+
+    Ok(())
+}
+
+#[derive(Debug)]
+pub struct CharFile(Arc<dyn CharDevice>);
+
+impl CharFile {
+    pub fn new(device: Arc<dyn CharDevice>) -> Self {
+        Self(device)
+    }
+}
+
+#[inherit_methods(from = "self.0.open().unwrap()")]
+impl InodeIo for CharFile {
+    fn read_at(
+        &self,
+        offset: usize,
+        writer: &mut VmWriter,
+        status_flags: StatusFlags,
+    ) -> Result<usize>;
+    fn write_at(
+        &self,
+        offset: usize,
+        reader: &mut VmReader,
+        status_flags: StatusFlags,
+    ) -> Result<usize>;
+}
+
+#[inherit_methods(from = "self.0.open().unwrap()")]
+impl Pollable for CharFile {
+    fn poll(&self, mask: IoEvents, poller: Option<&mut PollHandle>) -> IoEvents;
+}
+
+#[inherit_methods(from = "self.0.open().unwrap()")]
+impl FileIo for CharFile {
+    fn check_seekable(&self) -> Result<()>;
+    fn is_offset_aware(&self) -> bool;
+}
+
+impl Device for CharFile {
+    fn type_(&self) -> DeviceType {
+        DeviceType::Char
+    }
+
+    fn id(&self) -> DeviceId {
+        self.0.id()
+    }
+
+    fn open(&self) -> Result<Box<dyn FileIo>> {
+        Ok(Box::new(CharFile(self.0.clone())))
```

Comment:
Need to use two separate types to represent a char device inode and an _open_ char device file. Let's say one is `CharFile` and the other is `OpenCharFile`. `CharFile` need implement the `Device` trait, but only provide dummy or error-returning implementations for `InodeIo`, `Pollable`, and `FileIo`. On the other hand, `OpenCharFile` does not implement the `Device` trait, but implement the `InodeIo`, `Pollable`, and `FileIo` traits correctly.

The same `BlockFile` type merged in #2560 also suffers this problem. Fix it in a separate commit.